### PR TITLE
refactor: the long goodbye 2

### PR DESCRIPTION
**Pull request by dundargoc**
long is 32 bits on windows, while it is 64 bits on other architectures.
This makes the type suboptimal for a codebase meant to be
cross-platform. Replace it with more appropriate integer types.

### CODE REVIEW

**Comment by dundargoc** [comment]
Follow-up to https://github.com/neovim/neovim/pull/25417

File: src/nvim/eval/typval.c [hunk 1]
```diff
@@ -636,23 +636,23 @@ listitem_T *tv_list_check_range_index_one(list_T *const l, long *const n1, const
...
       return FAIL;
     }
-    *n2 = tv_list_idx_of_item(l, ni);
+    *n2 = (int)tv_list_idx_of_item(l, ni);
   }
 
   // Check that n2 isn't before n1.
   if (*n1 < 0) {
-    *n1 = tv_list_idx_of_item(l, li1);
+    *n1 = (int)tv_list_idx_of_item(l, li1);
```
Comment (zeertzjq) [inline src/nvim/eval/typval.c:655]: Also change return type of `tv_list_idx_of_item()` to `int`.

File: src/nvim/fileio.c [hunk 2]
```diff
@@ -3838,14 +3838,14 @@ long read_eintr(int fd, void *buf, size_t bufsize)
...
 {
-  long ret = 0;
+  int ret = 0;
 
   // Repeat the write() so long it didn't fail, other than being interrupted
   // by a signal.
-  while (ret < (long)bufsize) {
-    long wlen = write(fd, (char *)buf + ret, (unsigned)(bufsize - (size_t)ret));
+  while (ret < (int)bufsize) {
+    int wlen = (int)write(fd, (char *)buf + ret, (unsigned)(bufsize - (size_t)ret));
```
Comment (zeertzjq) [inline src/nvim/fileio.c:3848]: Maybe use `ssize_t` here?

File: src/nvim/buffer_defs.h [hunk 3]
```diff
@@ -440,10 +440,10 @@ struct file_buffer {
...
                                 // where invoked
 
-  long b_mtime;                 // last change time of original file
-  long b_mtime_ns;              // nanoseconds of last change time
-  long b_mtime_read;            // last change time when reading
-  long b_mtime_read_ns;         // nanoseconds of last read time
+  int b_mtime;                  // last change time of original file
+  int b_mtime_ns;               // nanoseconds of last change time
+  int b_mtime_read;             // last change time when reading
+  int b_mtime_read_ns;          // nanoseconds of last read time
```
Comment (zeertzjq) [inline src/nvim/buffer_defs.h:446]:
```suggestion
  int64_t b_mtime;              // last change time of original file
  int64_t b_mtime_ns;           // nanoseconds of last change time
  int64_t b_mtime_read;         // last change time when reading
  int64_t b_mtime_read_ns;      // nanoseconds of last read time
```

File: src/nvim/fileio.c [hunk 4]
```diff
@@ -2905,7 +2905,7 @@ int buf_check_timestamp(buf_T *buf)
       && (!(file_info_ok = os_fileinfo(buf->b_ffname, &file_info))
           || time_differs(&file_info, buf->b_mtime, buf->b_mtime_ns)
           || (int)file_info.stat.st_mode != buf->b_orig_mode)) {
-    const int prev_b_mtime = (int)buf->b_mtime;
+    const int prev_b_mtime = buf->b_mtime;
```
Comment (zeertzjq) [inline src/nvim/fileio.c:2908]:
```suggestion
    const int64_t prev_b_mtime = buf->b_mtime;
```

File: src/nvim/fileio.c [hunk 5]
```diff
@@ -3228,8 +3228,8 @@ void buf_reload(buf_T *buf, int orig_mode, bool reload_options)
 void buf_store_file_info(buf_T *buf, FileInfo *file_info)
   FUNC_ATTR_NONNULL_ALL
 {
-  buf->b_mtime = file_info->stat.st_mtim.tv_sec;
-  buf->b_mtime_ns = file_info->stat.st_mtim.tv_nsec;
+  buf->b_mtime = (int)file_info->stat.st_mtim.tv_sec;
+  buf->b_mtime_ns = (int)file_info->stat.st_mtim.tv_nsec;
```
Comment (zeertzjq) [inline src/nvim/fileio.c:3232]:
```suggestion
  buf->b_mtime = file_info->stat.st_mtim.tv_sec;
  buf->b_mtime_ns = file_info->stat.st_mtim.tv_nsec;
```

**Review by zeertzjq — DISMISSED** [review]

File: src/nvim/fileio.c [hunk 6]
```diff
@@ -2177,7 +2177,8 @@ void msg_add_lines(int insert_space, linenr_T lnum, off_T nchars)
   }
 }
 
-bool time_differs(const FileInfo *file_info, long mtime, long mtime_ns) FUNC_ATTR_CONST
+bool time_differs(const FileInfo *file_info, long mtime, long mtime_ns)
```
Comment (zeertzjq) [inline src/nvim/fileio.c:2180]:
```suggestion
bool time_differs(const FileInfo *file_info, int64_t mtime, int64_t mtime_ns)
```

File: src/nvim/bufwrite.c [hunk 7]
```diff
@@ -349,7 +349,7 @@ static int buf_write_bytes(struct bw_info *ip)
 static int check_mtime(buf_T *buf, FileInfo *file_info)
 {
   if (buf->b_mtime_read != 0
-      && time_differs(file_info, buf->b_mtime_read, buf->b_mtime_read_ns)) {
+      && time_differs(file_info, (long)buf->b_mtime_read, (long)buf->b_mtime_read_ns)) {
```
Comment (zeertzjq) [inline src/nvim/bufwrite.c:352]:
```suggestion
      && time_differs(file_info, buf->b_mtime_read, buf->b_mtime_read_ns)) {
```

File: src/nvim/fileio.c [hunk 8]
```diff
@@ -2903,9 +2904,9 @@ int buf_check_timestamp(buf_T *buf)
   if (!(buf->b_flags & BF_NOTEDITED)
       && buf->b_mtime != 0
       && (!(file_info_ok = os_fileinfo(buf->b_ffname, &file_info))
-          || time_differs(&file_info, buf->b_mtime, buf->b_mtime_ns)
+          || time_differs(&file_info, (long)buf->b_mtime, (long)buf->b_mtime_ns)
```
Comment (zeertzjq) [inline src/nvim/fileio.c:2907]:
```suggestion
          || time_differs(&file_info, buf->b_mtime, buf->b_mtime_ns)
```

File: src/nvim/eval.c [hunk 9]
```diff
@@ -6231,7 +6231,7 @@ void timer_due_cb(TimeWatcher *tw, void *data)
   timer_decref(timer);
 }
 
-uint64_t timer_start(const long timeout, const int repeat_count, const Callback *const callback)
+uint64_t timer_start(const int timeout, const int repeat_count, const Callback *const callback)
```
Comment (zeertzjq) [inline src/nvim/eval.c:6234]:
```suggestion
uint64_t timer_start(const int64_t timeout, const int repeat_count, const Callback *const callback)
```

File: src/nvim/eval.h [hunk 10]
```diff
@@ -225,7 +225,7 @@ typedef struct {
   int repeat_count;
   int refcount;
   int emsg_count;  ///< Errors in a repeating timer.
-  long timeout;
+  int timeout;
```
Comment (zeertzjq) [inline src/nvim/eval.h:228]:
```suggestion
  int64_t timeout;
```

File: src/nvim/eval/funcs.c [hunk 11]
```diff
@@ -8708,7 +8708,7 @@ static void f_timer_start(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
   if (!callback_from_typval(&callback, &argvars[1])) {
     return;
   }
-  rettv->vval.v_number = (varnumber_T)timer_start((const long)tv_get_number(&argvars[0]), repeat,
+  rettv->vval.v_number = (varnumber_T)timer_start((const int)tv_get_number(&argvars[0]), repeat,
                                                   &callback);
```
Comment (zeertzjq) [inline src/nvim/eval/funcs.c:8712]:
```suggestion
  rettv->vval.v_number = (varnumber_T)timer_start(tv_get_number(&argvars[0]), repeat, &callback);
```

File: src/nvim/diff.c [hunk 12]
```diff
@@ -3448,31 +3448,29 @@ static int parse_diff_ed(char *line, diffhunk_T *hunk)
...
   char *p = line;
   if (*p++ == '@' && *p++ == '@' && *p++ == ' ' && *p++ == '-') {
-    long oldcount;
-    long newline;
-    long newcount;
-    long oldline = getdigits_long(&p, true, 0);
+    int oldcount;
+    linenr_T newline;
+    int newcount;
+    linenr_T oldline = getdigits_int(&p, true, 0);
```
Comment (zeertzjq) [inline src/nvim/diff.c:3462]:
```suggestion
    linenr_T oldline = getdigits_int32(&p, true, 0);
```

File: src/nvim/fileio.c [hunk 13]
```diff
@@ -3823,12 +3824,12 @@ char *file_pat_to_reg_pat(const char *pat, const char *pat_end, char *allow_dirs
 
 /// Version of read() that retries when interrupted by EINTR (possibly
 /// by a SIGWINCH).
-long read_eintr(int fd, void *buf, size_t bufsize)
+int read_eintr(int fd, void *buf, size_t bufsize)
 {
-  long ret;
+  int ret;
```
Comment (zeertzjq) [inline src/nvim/fileio.c:3829]:
```suggestion
  ssize_t ret;
```

File: src/nvim/fileio.c [hunk 14]
```diff
@@ -3823,12 +3824,12 @@ char *file_pat_to_reg_pat(const char *pat, const char *pat_end, char *allow_dirs
...
 
   while (true) {
-    ret = read(fd, buf, (unsigned)bufsize);
+    ret = (int)read(fd, buf, (unsigned)bufsize);
```
Comment (zeertzjq) [inline src/nvim/fileio.c:3832]:
```suggestion
    ret = read(fd, buf, (unsigned)bufsize);
```

File: src/nvim/quickfix.c [hunk 15]
```diff
@@ -131,7 +131,7 @@ typedef struct qf_list_S {
   bool qf_multiline;
   bool qf_multiignore;
   bool qf_multiscan;
-  long qf_changedtick;
+  int qf_changedtick;
```
Comment (zeertzjq) [inline src/nvim/quickfix.c:134]:
```suggestion
  varnumber_T qf_changedtick;
```

File: src/nvim/quickfix.c [hunk 16]
```diff
@@ -2756,7 +2756,7 @@ static int qf_jump_edit_buffer(qf_info_T *qi, qfline_T *qf_ptr, int forceit, int
                                int *opened_window)
 {
   qf_list_T *qfl = qf_get_curlist(qi);
-  long old_changetick = qfl->qf_changedtick;
+  int old_changetick = qfl->qf_changedtick;
```
Comment (zeertzjq) [inline src/nvim/quickfix.c:2759]:
```suggestion
  varnumber_T old_changetick = qfl->qf_changedtick;
```

File: src/nvim/quickfix.c [hunk 17]
```diff
@@ -2891,7 +2891,7 @@ static void qf_jump_print_msg(qf_info_T *qi, int qf_index, qfline_T *qf_ptr, buf
 static int qf_jump_open_window(qf_info_T *qi, qfline_T *qf_ptr, bool newwin, int *opened_window)
 {
   qf_list_T *qfl = qf_get_curlist(qi);
-  long old_changetick = qfl->qf_changedtick;
+  int old_changetick = qfl->qf_changedtick;
```
Comment (zeertzjq) [inline src/nvim/quickfix.c:2894]:
```suggestion
  varnumber_T old_changetick = qfl->qf_changedtick;
```

**Review by bfredl — APPROVED** [review]
image

**Pull request merged**